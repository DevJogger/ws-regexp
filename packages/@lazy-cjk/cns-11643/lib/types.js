"use strict";
/**
 * Created by user on 2020/5/30.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumPinyinType = exports.EnumPinyinMode = void 0;
var EnumPinyinMode;
(function (EnumPinyinMode) {
    /**
     * 「CNS_pinyin_1」以調值(數字)呈現
     */
    EnumPinyinMode[EnumPinyinMode["num"] = 0] = "num";
    /**
     * 「CNS_pinyin_2」以聲調符號呈現
     * @type {EnumPinyinMode.tone}
     */
    EnumPinyinMode[EnumPinyinMode["tone"] = 1] = "tone";
})(EnumPinyinMode = exports.EnumPinyinMode || (exports.EnumPinyinMode = {}));
var EnumPinyinType;
(function (EnumPinyinType) {
    /**
     * 漢語(han)
     */
    EnumPinyinType[EnumPinyinType["han"] = 0] = "han";
    /**
     * 注音第二式(zuin2)
     */
    EnumPinyinType[EnumPinyinType["zuin2"] = 1] = "zuin2";
    /**
     * 耶魯(yale)
     */
    EnumPinyinType[EnumPinyinType["yale"] = 2] = "yale";
    /**
     * 韋式(wei)
     */
    EnumPinyinType[EnumPinyinType["wei"] = 3] = "wei";
})(EnumPinyinType = exports.EnumPinyinType || (exports.EnumPinyinType = {}));
//# sourceMappingURL=types.js.map