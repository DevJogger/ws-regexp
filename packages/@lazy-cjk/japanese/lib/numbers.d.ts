import { EnumTranscribeNumberConfigsKeys, IOptionsTranscribeNumber, IOptionsTranscribeNumberRuntime } from './types';
import { ITSTypeAndStringLiteral } from 'ts-type/lib/helper/string';
export * from './data/numbers';
export declare function handleTranscribeNumberOptions(config?: Partial<IOptionsTranscribeNumber> | ITSTypeAndStringLiteral<EnumTranscribeNumberConfigsKeys>): IOptionsTranscribeNumberRuntime;
/**
 * Unify input to string
 */
export declare function unifyInputToString(number: number | string): string;
export declare function transcribeNumber(number: number | string, inputConfig?: Partial<IOptionsTranscribeNumber> | ITSTypeAndStringLiteral<EnumTranscribeNumberConfigsKeys>): string;
export declare function _transcribeNumberCore(number: string, config: IOptionsTranscribeNumberRuntime, deep?: number): string;
export default transcribeNumber;
