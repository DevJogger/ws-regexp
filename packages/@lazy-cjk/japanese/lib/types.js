"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumTranscribeNumberConfigsKeys = exports.EnumRomanizationConfigsKeys = void 0;
var EnumRomanizationConfigsKeys;
(function (EnumRomanizationConfigsKeys) {
    EnumRomanizationConfigsKeys["wikipedia"] = "wikipedia";
    EnumRomanizationConfigsKeys["traditional_hepburn"] = "traditional hepburn";
    EnumRomanizationConfigsKeys["modified_hepburn"] = "modified hepburn";
    EnumRomanizationConfigsKeys["kunrei"] = "kunrei";
    EnumRomanizationConfigsKeys["nihon"] = "nihon";
})(EnumRomanizationConfigsKeys || (exports.EnumRomanizationConfigsKeys = EnumRomanizationConfigsKeys = {}));
var EnumTranscribeNumberConfigsKeys;
(function (EnumTranscribeNumberConfigsKeys) {
    EnumTranscribeNumberConfigsKeys["default"] = "default";
    EnumTranscribeNumberConfigsKeys["formal"] = "formal";
    EnumTranscribeNumberConfigsKeys["traditional"] = "traditional";
})(EnumTranscribeNumberConfigsKeys || (exports.EnumTranscribeNumberConfigsKeys = EnumTranscribeNumberConfigsKeys = {}));
//# sourceMappingURL=types.js.map