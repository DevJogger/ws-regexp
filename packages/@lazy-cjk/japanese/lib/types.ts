/**
 * Created by user on 2020/5/31.
 */
import { ITSTypeAndStringLiteral } from 'ts-type/lib/helper/string';
import { predefineedTranscriptionConfigs } from './numbers';
import { ITSOverwrite } from 'ts-type/lib/type/record';

export const enum EnumRomanizationConfigsKeys
{
	'wikipedia' = 'wikipedia',
	'traditional_hepburn' = 'traditional hepburn',
	'modified_hepburn' = 'modified hepburn',
	'kunrei' = 'kunrei',
	'nihon' = 'nihon',
}

export type IRomanizationConfigsKeys = ITSTypeAndStringLiteral<EnumRomanizationConfigsKeys>;

/**
 * Config is represented as plain object,
 * where object keys stand for a collection of similar characters,
 * and the value determines how these characters are converted.
 * So the object is not just the same as a conversion table.
 */
export interface IOptionsRomanize
{
	し?: string;
	ち?: string;
	つ?: string;
	ふ?: string;
	じ?: string;
	ぢ?: string;
	づ?: string;
	ああ?: string;
	いい?: string;
	うう?: string;
	ええ?: string;
	おお?: string;
	あー?: string;
	えい?: string;
	おう?: string;
	んあ?: string;
	んば?: string;
	っち?: string;
	ゐ?: string;
	を?: string;
	punctuation?: boolean;
	configPreset?: IRomanizationConfigsKeys;
	ignoreUnSupported?: boolean,
}

export const enum EnumTranscribeNumberConfigsKeys
{
	'default' = 'default',
	formal = 'formal',
	traditional = 'traditional',
}

type IHelperPredefineedTranscriptionConfigsSub2<K extends keyof typeof predefineedTranscriptionConfigs> =
	typeof predefineedTranscriptionConfigs[K][keyof typeof predefineedTranscriptionConfigs[K]]

type IHelperPredefineedTranscriptionConfigsSub<K extends keyof typeof predefineedTranscriptionConfigs> =
	keyof typeof predefineedTranscriptionConfigs[K]
	| IHelperPredefineedTranscriptionConfigsSub2<K>

export interface ITranscribeNumberConfig
{
	unitNames?: IHelperPredefineedTranscriptionConfigsSub<"unitNames">,

	digits: IHelperPredefineedTranscriptionConfigsSub<"digits">,
	specialUnitNames: IHelperPredefineedTranscriptionConfigsSub<"specialUnitNames">,
	smallUnitNames: IHelperPredefineedTranscriptionConfigsSub<"smallUnitNames">,

	truncateOne?: string[],
	minusSign?: string,
	decimalPoint?: string,
}

export interface IOptionsTranscribeNumber extends ITranscribeNumberConfig
{
	configPreset?: ITSTypeAndStringLiteral<EnumTranscribeNumberConfigsKeys>;

	/**
	 * Get sanitized unit name keys
	 */
	keysOfUnitNames?: number[]
}

export interface IOptionsTranscribeNumberRuntime extends ITSOverwrite<IOptionsTranscribeNumber, {
	unitNames?: IHelperPredefineedTranscriptionConfigsSub2<"unitNames">,

	digits: IHelperPredefineedTranscriptionConfigsSub2<"digits">,
	specialUnitNames: IHelperPredefineedTranscriptionConfigsSub2<"specialUnitNames">,
	smallUnitNames: IHelperPredefineedTranscriptionConfigsSub2<"smallUnitNames">,
}>
{

}
