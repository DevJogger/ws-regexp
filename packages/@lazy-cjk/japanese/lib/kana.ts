import { specialKatakanizationTable, specialHiraganizationTable, hiraganaRegex, katakanaRegex } from './data/kana';

export * from './data/kana';

const chr = String.fromCharCode;
const ord = function (char: string)
{
	return char.charCodeAt(0);
};

/**
 * Convert input katakana into hiragana.
 */
export function hiraganize(string: string)
{
	return string.replace(katakanaRegex, function (katakana)
	{
		if (katakana.match(/^[\u30a1-\u30f4\u30fd\u30fe]$/))
		{
			return chr(ord(katakana) - ord('ァ') + ord('ぁ'));
		}
		else if (specialHiraganizationTable[katakana])
		{
			return specialHiraganizationTable[katakana];
		}
	});
}

/**
 * Convert input hiragana into katakana.
 */
export function katakanize(string: string)
{
	return string.replace(hiraganaRegex, function (hiragana)
	{
		if (hiragana.match(/^[\u3041-\u3094\u309d\u309e]$/))
		{
			return chr(ord(hiragana) - ord('ぁ') + ord('ァ'));
		}
		else if (specialKatakanizationTable[hiragana])
		{
			return specialKatakanizationTable[hiragana];
		}
	});
}

