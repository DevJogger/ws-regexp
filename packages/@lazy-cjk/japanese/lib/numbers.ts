'use strict';

import extend from 'lodash/defaults';
import Big from 'big.js';
import {
	predefineedTranscriptionConfigs,
	transcriptionConfigs,
	MIN_SAFE_INTEGER,
	MAX_SAFE_INTEGER,
} from './data/numbers';
import { getBits, getBit } from './util/numbers';

export * from './data/numbers';

export function transcribeNumber(number: number | string, config?): string
{
	if (typeof config === 'undefined')
	{
		// default config
		config = transcriptionConfigs['default'];
	}

	if (typeof config === 'string')
	{
		config = transcriptionConfigs[config];

		if (typeof config === 'undefined')
		{
			throw new ReferenceError('Transcription method "' + config + '" is undefined');
		}
	}

	if (typeof config === 'object')
	{
		config = extend({}, config, transcriptionConfigs['default']);
	}
	else
	{
		throw new Error('You specified unknown config to japanese.transcribeNumber');
	}

	if (typeof config.digits === 'string')
	{
		config.digits = predefineedTranscriptionConfigs.digits[config.digits];

		if (typeof config.digits === 'undefined')
		{
			throw new ReferenceError('Transcription method of digits "' + config.digits + '" is undefined');
		}
	}

	if (typeof config.unitNames === 'string')
	{
		config.unitNames = predefineedTranscriptionConfigs.unitNames[config.unitNames];

		if (typeof config.unitNames === 'undefined')
		{
			throw new ReferenceError('Transcription method of unitNames "' + config.unitNames + '" is undefined');
		}
	}

	if (typeof config.specialUnitNames === 'string')
	{
		config.specialUnitNames = predefineedTranscriptionConfigs.specialUnitNames[config.specialUnitNames];

		if (typeof config.specialUnitNames === 'undefined')
		{
			throw new ReferenceError('Transcription method of specialUnitNames "' + config.specialUnitNames + '" is undefined');
		}
	}

	if (typeof config.smallUnitNames === 'string')
	{
		config.smallUnitNames = predefineedTranscriptionConfigs.smallUnitNames[config.smallUnitNames];

		if (typeof config.smallUnitNames === 'undefined')
		{
			throw new ReferenceError('Transcription method of smallUnitNames "' + config.smallUnitNames + '" is undefined');
		}
	}

	// Unify input to string

	if (typeof number === 'number')
	{
		if (MIN_SAFE_INTEGER <= number && number < MAX_SAFE_INTEGER)
		{
			number = number.toString();
		}
		else
		{
			// Paste number into binary form
			const buf = Buffer.alloc(8);
			buf.writeDoubleBE(number, 0);

			let sign = getBit(buf, 0);
			let exponent = getBits(buf, 1, 11) as any as number;
			let mantissa = getBits(buf, 12, 52);
			let fraction: Big = null;

			exponent = parseInt(exponent.toString());

			if (exponent === 0)
			{
				fraction = mantissa;
				exponent = 1;
			}
			else
			{
				fraction = (new Big(2)).pow(52).plus(mantissa);
			}

			number = fraction.times((new Big(2)).pow(exponent - 1023 - 52)).toFixed();

			if (sign)
			{
				number = '-' + number;
			}
		}
	}
	else if (typeof number !== 'string')
	{
		throw new ReferenceError('Type of `number` is unsupported');
	}

	let length = number.length;

	// Main convertion starts here

	let lit = '';
	let restoreZero = false;
	if (config.unitNames.lit && length > config.unitNames.lit)
	{
		lit = number.slice(0, -config.unitNames.lit).split('').map(function (digit)
		{
			return config.digits[digit];
		}).join('');

		number = number.slice(-config.unitNames.lit);
		length = number.length;
		if (number[0] === '0')
		{
			restoreZero = true;
			number = '9' + number.slice(1);
		}
	}

	// handle zero
	if (number === '0')
	{
		return config.digits[0];
	}

	let transcription = '';

	if (number.slice(-1) !== '0')
	{
		transcription += config.digits[(number as string).slice(-1)];
	}

	// Get sanitized unit name keys
	let keysOfUnitNames = Object.keys(config.unitNames).map(function (key)
	{
		// convert to int
		return parseInt(key);
	}).filter(function (key, index, self)
	{
		// unique
		return self.indexOf(key) === index;
	}).filter(function (key)
	{
		// validate
		return isFinite(key) && key > 0;
	}).sort(function (a, b)
	{
		// asc sort
		return a - b;
	});

	keysOfUnitNames.forEach(function (key, index)
	{
		let nextKey = keysOfUnitNames[index + 1] || Infinity;
		// slice the digits spaned by the unit name
		let token = (number as string).slice(Math.max(length - nextKey, 0), Math.max(length - key, 0));

		if (token.length > 0)
		{
			// check if every number in the token is zero
			if (!token.split('').every(function (digit) { return digit === '0'; }))
			{
				// truncateOne
				if (config.truncateOne.indexOf(config.unitNames[key]) !== -1 && parseInt(token) === 1)
				{
					transcription = config.unitNames[key] + transcription;
				}
				else
				{
					transcription = transcribeNumber(token, config) + config.unitNames[key] + transcription;
				}
			}
		}
	});

	// Rejoin lit tokens
	if (restoreZero)
	{
		transcription = transcription.replace(new RegExp('^' + config.digits[9]), config.digits[0]);
	}
	transcription = lit + transcription;

	return transcription;
}

export default transcribeNumber
