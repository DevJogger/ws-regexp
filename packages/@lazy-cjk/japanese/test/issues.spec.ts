//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { basename, extname } from 'path';
import { handleRomanizeOptionsAndTable, romanize } from '../lib/romanize';
import { EnumRomanizationConfigsKeys } from '../lib/types';

beforeAll(async () =>
{

});

let word = `不支援 𠮷𠬠𡬶𫗭𣛙𢎐 ...這類字🍔\t🍕\t🍖\t🍗\t🍘\t🍙\t🍚\t🍛[Pixiv] Judo ユフォリア (5026941)(呉マサヒロ)] 高飛車な生徒会長をアプリで従順調教 ☆﹐﹑veevee˚ㆍ ㅔㅣ ㅑㅣㅗ므ㅛㅕㅇㅁ ㅇㅜㅇ ㅏㅏ Whiskers ˶˃ᆺ˂˶`;

describe(basename(__filename, extname(__filename)), () =>
{

	test.skip(`dummy`, () => {});

	// @see https://github.com/bluelovers/ws-regexp/issues/2
	test(`ignoreUnSupported`, () =>
	{

		let actual = romanize(word, {
			ignoreUnSupported: true,
		});

		expect(actual).toMatchSnapshot();

	});

	test(`configPreset`, () =>
	{
		expect(handleRomanizeOptionsAndTable()).toMatchSnapshot();

		expect(handleRomanizeOptionsAndTable(EnumRomanizationConfigsKeys.traditional_hepburn)).toMatchSnapshot();

		expect(handleRomanizeOptionsAndTable({
			configPreset: EnumRomanizationConfigsKeys.traditional_hepburn,
		})).toMatchSnapshot();
	});

})
