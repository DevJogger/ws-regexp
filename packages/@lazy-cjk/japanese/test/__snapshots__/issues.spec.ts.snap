// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issues.spec configPreset 1`] = `
{
  "config": {
    "punctuation": true,
    "ああ": "aa",
    "あー": "ā",
    "いい": "ii",
    "うう": "ū",
    "えい": "ei",
    "ええ": "ee",
    "おう": "ō",
    "おお": "ō",
    "し": "shi",
    "じ": "ji",
    "ち": "chi",
    "ぢ": "ji",
    "っち": "tchi",
    "つ": "tsu",
    "づ": "zu",
    "ふ": "fu",
    "ゐ": "i",
    "を": "o",
    "んあ": "n'a",
    "んば": "nba",
  },
  "table": {
    "ぁ": "a",
    "あ": "a",
    "ぃ": "i",
    "い": "i",
    "いぃ": "yi",
    "いぇ": "ye",
    "ぅ": "u",
    "う": "u",
    "うぁ": "wa",
    "うぃ": "wi",
    "うぅ": "wu",
    "うぇ": "we",
    "うぉ": "wo",
    "うゅ": "wyu",
    "ぇ": "e",
    "え": "e",
    "ぉ": "o",
    "お": "o",
    "か": "ka",
    "が": "ga",
    "き": "ki",
    "きぇ": "kye",
    "きゃ": "kya",
    "きゅ": "kyu",
    "きょ": "kyo",
    "ぎ": "gi",
    "ぎぇ": "gye",
    "ぎゃ": "gya",
    "ぎゅ": "gyu",
    "ぎょ": "gyo",
    "く": "ku",
    "くぁ": "kwa",
    "くぃ": "kwi",
    "くぅ": "kwu",
    "くぇ": "kwe",
    "くぉ": "kwo",
    "くゎ": "kwa",
    "ぐ": "gu",
    "ぐぁ": "gwa",
    "ぐぃ": "gwi",
    "ぐぅ": "gwu",
    "ぐぇ": "gwe",
    "ぐぉ": "gwo",
    "ぐゎ": "gwa",
    "け": "ke",
    "げ": "ge",
    "こ": "ko",
    "ご": "go",
    "さ": "sa",
    "ざ": "za",
    "し": "shi",
    "しぇ": "she",
    "しゃ": "sha",
    "しゅ": "shu",
    "しょ": "sho",
    "じ": "ji",
    "じぇ": "je",
    "じゃ": "ja",
    "じゅ": "ju",
    "じょ": "jo",
    "す": "su",
    "すぃ": "si",
    "ず": "zu",
    "ずぃ": "zi",
    "せ": "se",
    "ぜ": "ze",
    "そ": "so",
    "ぞ": "zo",
    "た": "ta",
    "だ": "da",
    "ち": "chi",
    "ちぇ": "che",
    "ちゃ": "cha",
    "ちゅ": "chu",
    "ちょ": "cho",
    "ぢ": "ji",
    "ぢぇ": "je",
    "ぢゃ": "ja",
    "ぢゅ": "ju",
    "ぢょ": "jo",
    "っ": "tu",
    "つ": "tsu",
    "つぁ": "tsa",
    "つぃ": "tsi",
    "つぇ": "tse",
    "つぉ": "tso",
    "つゅ": "tsyu",
    "づ": "zu",
    "づぁ": "zua",
    "づぃ": "zui",
    "づぇ": "zue",
    "づぉ": "zuo",
    "て": "te",
    "てぃ": "ti",
    "てぇ": "tee",
    "てゃ": "tea",
    "てゅ": "tyu",
    "てょ": "teo",
    "で": "de",
    "でぃ": "di",
    "でぇ": "dee",
    "でゃ": "dea",
    "でゅ": "dyu",
    "でょ": "deo",
    "と": "to",
    "とぅ": "tu",
    "ど": "do",
    "どぅ": "du",
    "な": "na",
    "に": "ni",
    "にぇ": "nye",
    "にゃ": "nya",
    "にゅ": "nyu",
    "にょ": "nyo",
    "ぬ": "nu",
    "ね": "ne",
    "の": "no",
    "は": "ha",
    "ば": "ba",
    "ぱ": "pa",
    "ひ": "hi",
    "ひぇ": "hye",
    "ひゃ": "hya",
    "ひゅ": "hyu",
    "ひょ": "hyo",
    "び": "bi",
    "びぇ": "bye",
    "びゃ": "bya",
    "びゅ": "byu",
    "びょ": "byo",
    "ぴ": "pi",
    "ぴぇ": "pye",
    "ぴゃ": "pya",
    "ぴゅ": "pyu",
    "ぴょ": "pyo",
    "ふ": "fu",
    "ふぁ": "fa",
    "ふぃ": "fi",
    "ふぇ": "fe",
    "ふぉ": "fo",
    "ふゃ": "fya",
    "ふゅ": "fyu",
    "ふょ": "fyo",
    "ぶ": "bu",
    "ぷ": "pu",
    "へ": "he",
    "べ": "be",
    "ぺ": "pe",
    "ほ": "ho",
    "ほぅ": "hu",
    "ぼ": "bo",
    "ぽ": "po",
    "ま": "ma",
    "み": "mi",
    "みぇ": "mye",
    "みゃ": "mya",
    "みゅ": "my",
    "みょ": "myo",
    "む": "mu",
    "め": "me",
    "も": "mo",
    "ゃ": "ya",
    "や": "ya",
    "ゅ": "yu",
    "ゆ": "yu",
    "ょ": "yo",
    "よ": "yo",
    "ら": "ra",
    "ら゚": "la",
    "り": "ri",
    "りぇ": "rye",
    "りゃ": "rya",
    "りゅ": "ryu",
    "りょ": "ryo",
    "り゚": "li",
    "る": "ru",
    "る゚": "lu",
    "れ": "re",
    "れ゚": "le",
    "ろ": "ro",
    "ろ゚": "lo",
    "ゎ": "wa",
    "わ": "wa",
    "わ゙": "va",
    "ゐ": "i",
    "ゐ゙": "vi",
    "ゑ": "e",
    "ゑ゙": "ve",
    "を": "o",
    "を゙": "vo",
    "ん": "n",
    "ゔ": "vu",
    "ゔぁ": "va",
    "ゔぃ": "vi",
    "ゔぇ": "ve",
    "ゔぉ": "vo",
    "ゔゃ": "vya",
    "ゔゅ": "vyu",
    "ゔょ": "vyo",
    "ヵ": "ka",
    "ヶ": "ke",
  },
}
`;

exports[`issues.spec configPreset 2`] = `
{
  "config": {
    "punctuation": true,
    "ああ": "aa",
    "あー": "ā",
    "いい": "ii",
    "うう": "ū",
    "えい": "ei",
    "ええ": "ee",
    "おう": "ō",
    "おお": "ō",
    "し": "shi",
    "じ": "ji",
    "ち": "chi",
    "ぢ": "ji",
    "っち": "tchi",
    "つ": "tsu",
    "づ": "zu",
    "ふ": "fu",
    "ゐ": "i",
    "を": "wo",
    "んあ": "n-a",
    "んば": "mba",
  },
  "table": {
    "ぁ": "a",
    "あ": "a",
    "ぃ": "i",
    "い": "i",
    "いぃ": "yi",
    "いぇ": "ye",
    "ぅ": "u",
    "う": "u",
    "うぁ": "wa",
    "うぃ": "wi",
    "うぅ": "wu",
    "うぇ": "we",
    "うぉ": "wo",
    "うゅ": "wyu",
    "ぇ": "e",
    "え": "e",
    "ぉ": "o",
    "お": "o",
    "か": "ka",
    "が": "ga",
    "き": "ki",
    "きぇ": "kye",
    "きゃ": "kya",
    "きゅ": "kyu",
    "きょ": "kyo",
    "ぎ": "gi",
    "ぎぇ": "gye",
    "ぎゃ": "gya",
    "ぎゅ": "gyu",
    "ぎょ": "gyo",
    "く": "ku",
    "くぁ": "kwa",
    "くぃ": "kwi",
    "くぅ": "kwu",
    "くぇ": "kwe",
    "くぉ": "kwo",
    "くゎ": "kwa",
    "ぐ": "gu",
    "ぐぁ": "gwa",
    "ぐぃ": "gwi",
    "ぐぅ": "gwu",
    "ぐぇ": "gwe",
    "ぐぉ": "gwo",
    "ぐゎ": "gwa",
    "け": "ke",
    "げ": "ge",
    "こ": "ko",
    "ご": "go",
    "さ": "sa",
    "ざ": "za",
    "し": "shi",
    "しぇ": "she",
    "しゃ": "sha",
    "しゅ": "shu",
    "しょ": "sho",
    "じ": "ji",
    "じぇ": "je",
    "じゃ": "ja",
    "じゅ": "ju",
    "じょ": "jo",
    "す": "su",
    "すぃ": "si",
    "ず": "zu",
    "ずぃ": "zi",
    "せ": "se",
    "ぜ": "ze",
    "そ": "so",
    "ぞ": "zo",
    "た": "ta",
    "だ": "da",
    "ち": "chi",
    "ちぇ": "che",
    "ちゃ": "cha",
    "ちゅ": "chu",
    "ちょ": "cho",
    "ぢ": "ji",
    "ぢぇ": "je",
    "ぢゃ": "ja",
    "ぢゅ": "ju",
    "ぢょ": "jo",
    "っ": "tu",
    "つ": "tsu",
    "つぁ": "tsa",
    "つぃ": "tsi",
    "つぇ": "tse",
    "つぉ": "tso",
    "つゅ": "tsyu",
    "づ": "zu",
    "づぁ": "zua",
    "づぃ": "zui",
    "づぇ": "zue",
    "づぉ": "zuo",
    "て": "te",
    "てぃ": "ti",
    "てぇ": "tee",
    "てゃ": "tea",
    "てゅ": "tyu",
    "てょ": "teo",
    "で": "de",
    "でぃ": "di",
    "でぇ": "dee",
    "でゃ": "dea",
    "でゅ": "dyu",
    "でょ": "deo",
    "と": "to",
    "とぅ": "tu",
    "ど": "do",
    "どぅ": "du",
    "な": "na",
    "に": "ni",
    "にぇ": "nye",
    "にゃ": "nya",
    "にゅ": "nyu",
    "にょ": "nyo",
    "ぬ": "nu",
    "ね": "ne",
    "の": "no",
    "は": "ha",
    "ば": "ba",
    "ぱ": "pa",
    "ひ": "hi",
    "ひぇ": "hye",
    "ひゃ": "hya",
    "ひゅ": "hyu",
    "ひょ": "hyo",
    "び": "bi",
    "びぇ": "bye",
    "びゃ": "bya",
    "びゅ": "byu",
    "びょ": "byo",
    "ぴ": "pi",
    "ぴぇ": "pye",
    "ぴゃ": "pya",
    "ぴゅ": "pyu",
    "ぴょ": "pyo",
    "ふ": "fu",
    "ふぁ": "fa",
    "ふぃ": "fi",
    "ふぇ": "fe",
    "ふぉ": "fo",
    "ふゃ": "fya",
    "ふゅ": "fyu",
    "ふょ": "fyo",
    "ぶ": "bu",
    "ぷ": "pu",
    "へ": "he",
    "べ": "be",
    "ぺ": "pe",
    "ほ": "ho",
    "ほぅ": "hu",
    "ぼ": "bo",
    "ぽ": "po",
    "ま": "ma",
    "み": "mi",
    "みぇ": "mye",
    "みゃ": "mya",
    "みゅ": "my",
    "みょ": "myo",
    "む": "mu",
    "め": "me",
    "も": "mo",
    "ゃ": "ya",
    "や": "ya",
    "ゅ": "yu",
    "ゆ": "yu",
    "ょ": "yo",
    "よ": "yo",
    "ら": "ra",
    "ら゚": "la",
    "り": "ri",
    "りぇ": "rye",
    "りゃ": "rya",
    "りゅ": "ryu",
    "りょ": "ryo",
    "り゚": "li",
    "る": "ru",
    "る゚": "lu",
    "れ": "re",
    "れ゚": "le",
    "ろ": "ro",
    "ろ゚": "lo",
    "ゎ": "wa",
    "わ": "wa",
    "わ゙": "va",
    "ゐ": "i",
    "ゐ゙": "vi",
    "ゑ": "e",
    "ゑ゙": "ve",
    "を": "wo",
    "を゙": "vo",
    "ん": "n",
    "ゔ": "vu",
    "ゔぁ": "va",
    "ゔぃ": "vi",
    "ゔぇ": "ve",
    "ゔぉ": "vo",
    "ゔゃ": "vya",
    "ゔゅ": "vyu",
    "ゔょ": "vyo",
    "ヵ": "ka",
    "ヶ": "ke",
  },
}
`;

exports[`issues.spec configPreset 3`] = `
{
  "config": {
    "configPreset": "traditional hepburn",
    "を": "wo",
    "んあ": "n-a",
    "んば": "mba",
  },
  "table": {
    "ぁ": "a",
    "あ": "a",
    "ぃ": "i",
    "い": "i",
    "いぃ": "yi",
    "いぇ": "ye",
    "ぅ": "u",
    "う": "u",
    "うぁ": "wa",
    "うぃ": "wi",
    "うぅ": "wu",
    "うぇ": "we",
    "うぉ": "wo",
    "うゅ": "wyu",
    "ぇ": "e",
    "え": "e",
    "ぉ": "o",
    "お": "o",
    "か": "ka",
    "が": "ga",
    "き": "ki",
    "きぇ": "kye",
    "きゃ": "kya",
    "きゅ": "kyu",
    "きょ": "kyo",
    "ぎ": "gi",
    "ぎぇ": "gye",
    "ぎゃ": "gya",
    "ぎゅ": "gyu",
    "ぎょ": "gyo",
    "く": "ku",
    "くぁ": "kwa",
    "くぃ": "kwi",
    "くぅ": "kwu",
    "くぇ": "kwe",
    "くぉ": "kwo",
    "くゎ": "kwa",
    "ぐ": "gu",
    "ぐぁ": "gwa",
    "ぐぃ": "gwi",
    "ぐぅ": "gwu",
    "ぐぇ": "gwe",
    "ぐぉ": "gwo",
    "ぐゎ": "gwa",
    "け": "ke",
    "げ": "ge",
    "こ": "ko",
    "ご": "go",
    "さ": "sa",
    "ざ": "za",
    "し": "si",
    "しぇ": "sye",
    "しゃ": "sya",
    "しゅ": "syu",
    "しょ": "syo",
    "じ": "zi",
    "じぇ": "zye",
    "じゃ": "zya",
    "じゅ": "zyu",
    "じょ": "zyo",
    "す": "su",
    "すぃ": "si",
    "ず": "zu",
    "ずぃ": "zi",
    "せ": "se",
    "ぜ": "ze",
    "そ": "so",
    "ぞ": "zo",
    "た": "ta",
    "だ": "da",
    "ち": "ti",
    "ちぇ": "tye",
    "ちゃ": "tya",
    "ちゅ": "tyu",
    "ちょ": "tyo",
    "ぢ": "di",
    "ぢぇ": "dye",
    "ぢゃ": "dya",
    "ぢゅ": "dyu",
    "ぢょ": "dyo",
    "っ": "tu",
    "つ": "tu",
    "つぁ": "tua",
    "つぃ": "tui",
    "つぇ": "tue",
    "つぉ": "tuo",
    "つゅ": "tuyu",
    "づ": "du",
    "づぁ": "dua",
    "づぃ": "dui",
    "づぇ": "due",
    "づぉ": "duo",
    "て": "te",
    "てぃ": "tei",
    "てぇ": "tee",
    "てゃ": "tea",
    "てゅ": "teu",
    "てょ": "teo",
    "で": "de",
    "でぃ": "dei",
    "でぇ": "dee",
    "でゃ": "dea",
    "でゅ": "deu",
    "でょ": "deo",
    "と": "to",
    "とぅ": "tou",
    "ど": "do",
    "どぅ": "dou",
    "な": "na",
    "に": "ni",
    "にぇ": "nye",
    "にゃ": "nya",
    "にゅ": "nyu",
    "にょ": "nyo",
    "ぬ": "nu",
    "ね": "ne",
    "の": "no",
    "は": "ha",
    "ば": "ba",
    "ぱ": "pa",
    "ひ": "hi",
    "ひぇ": "hye",
    "ひゃ": "hya",
    "ひゅ": "hyu",
    "ひょ": "hyo",
    "び": "bi",
    "びぇ": "bye",
    "びゃ": "bya",
    "びゅ": "byu",
    "びょ": "byo",
    "ぴ": "pi",
    "ぴぇ": "pye",
    "ぴゃ": "pya",
    "ぴゅ": "pyu",
    "ぴょ": "pyo",
    "ふ": "hu",
    "ふぁ": "hua",
    "ふぃ": "hui",
    "ふぇ": "hue",
    "ふぉ": "huo",
    "ふゃ": "huya",
    "ふゅ": "huyu",
    "ふょ": "huyo",
    "ぶ": "bu",
    "ぷ": "pu",
    "へ": "he",
    "べ": "be",
    "ぺ": "pe",
    "ほ": "ho",
    "ほぅ": "hu",
    "ぼ": "bo",
    "ぽ": "po",
    "ま": "ma",
    "み": "mi",
    "みぇ": "mye",
    "みゃ": "mya",
    "みゅ": "my",
    "みょ": "myo",
    "む": "mu",
    "め": "me",
    "も": "mo",
    "ゃ": "ya",
    "や": "ya",
    "ゅ": "yu",
    "ゆ": "yu",
    "ょ": "yo",
    "よ": "yo",
    "ら": "ra",
    "ら゚": "la",
    "り": "ri",
    "りぇ": "rye",
    "りゃ": "rya",
    "りゅ": "ryu",
    "りょ": "ryo",
    "り゚": "li",
    "る": "ru",
    "る゚": "lu",
    "れ": "re",
    "れ゚": "le",
    "ろ": "ro",
    "ろ゚": "lo",
    "ゎ": "wa",
    "わ": "wa",
    "わ゙": "va",
    "ゐ": "wi",
    "ゐ゙": "vi",
    "ゑ": "we",
    "ゑ゙": "ve",
    "を": "wo",
    "を゙": "vo",
    "ん": "n",
    "ゔ": "vu",
    "ゔぁ": "va",
    "ゔぃ": "vi",
    "ゔぇ": "ve",
    "ゔぉ": "vo",
    "ゔゃ": "vya",
    "ゔゅ": "vyu",
    "ゔょ": "vyo",
    "ヵ": "ka",
    "ヶ": "ke",
  },
}
`;

exports[`issues.spec ignoreUnSupported 1`] = `"不支援 𠮷𠬠𡬶𫗭𣛙𢎐 ...這類字🍔	🍕	🍖	🍗	🍘	🍙	🍚	🍛[Pixiv] Judo yuforia (5026941)(呉masahiro)] 高飛車na生徒会長oapuride従順調教 ☆﹐﹑veevee˚ㆍ ㅔㅣ ㅑㅣㅗ므ㅛㅕㅇㅁ ㅇㅜㅇ ㅏㅏ Whiskers ˶˃ᆺ˂˶"`;
